desc/objectives/_coils.py

```diff
@@ -708,6 +716,362 @@ def compute(self, params, constants=None):
+            self._eq_data_keys,
+            obj=eq,
+            grid=plasma_grid,
+            has_axis=plasma_grid.axis.size,
+        )
+        eq_transforms = get_transforms(
+            self._eq_data_keys,
+            obj=eq,
+            grid=plasma_grid,
+            has_axis=plasma_grid.axis.size,
+        )
+
+        self._constants = {
+            "coilset": coilset,
+            "coil_grid": coil_grid,
+            "eq_profiles": eq_profiles,
+            "eq_transforms": eq_transforms,
+            "quad_weights": 1.0,
+        }
+
+        if self._eq_fixed:
+            # precompute the equilibrium surface coordinates
+            data = compute_fun(
+                eq,
+                self._eq_data_keys,
+                params=eq.params_dict,
+                transforms=eq_transforms,
+                profiles=eq_profiles,
+            )
+            plasma_coords = rpz2xyz(jnp.array([data["R"], data["phi"], data["Z"]]).T)
+            self._constants["plasma_coords"] = plasma_coords
+
+        if self._normalize:
+            scales = compute_scaling_factors(eq)
+            self._normalization = scales["a"]
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, coils_params, eq_params=None, constants=None):
+        """Compute minimum distance between coils and the plasma/surface.
+
+        Parameters
+        ----------
+        coils_params : dict
+            Dictionary of coilset degrees of freedom, eg ``CoilSet.params_dict``
+        eq_params : dict
+            Dictionary of equilibrium or surface degrees of freedom,
+            eg ``Equilibrium.params_dict``
+            Only required if ``self._eq_fixed = False``.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc.
+            Defaults to self._constants.
+
+        Returns
+        -------
+        f : array of floats
+            Minimum distance from coil to surface for each coil in the coilset.
+
+        """
+        if constants is None:
+            constants = self.constants
+
+        # coil pts; shape(ncoils,coils_grid.num_nodes,3)
+        coils_pts = constants["coilset"]._compute_position(
+            params=coils_params, grid=constants["coil_grid"]
+        )
+
+        # plasma pts; shape(plasma_grid.num_nodes,3)
+        if self._eq_fixed:
+            plasma_pts = constants["plasma_coords"]
+        else:
+            data = compute_fun(
+                self._eq,
+                self._eq_data_keys,
+                params=eq_params,
+                transforms=constants["eq_transforms"],
+                profiles=constants["eq_profiles"],
+            )
+            plasma_pts = rpz2xyz(jnp.array([data["R"], data["phi"], data["Z"]]).T)
+
+        def body(k):
+            # dist btwn all pts; shape(ncoils,plasma_grid.num_nodes,coil_grid.num_nodes)
+            dist = safenorm(coils_pts[k][None, :, :] - plasma_pts[:, None, :], axis=-1)
```

Comment:
I think this could be useful for umbilic, where I want to find the distance between a curve and surface.